File: DapperDino.DAL/Models/Ticket.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text;

namespace DapperDino.DAL.Models
{
    public class Ticket:IEntity
    {
        public string Description { get; set; }
        [Required]
        public string Subject { get; set; }

        public int ApplicantId { get; set; }

        public int Priority { get; set; } = 0;
        public TicketStatus Status { get; set; } = TicketStatus.Open;
        public TicketCategory Category { get; set; } = TicketCategory.Undecided;

        public DateTime CreatedOn { get; set; } = DateTime.Now;
        public DateTime LastModified { get; set; } = DateTime.Now;

        public int? LanguageId { get; set; }
        [ForeignKey("LanguageId")]
        public virtual Proficiency Language { get; set; }

        public int? FrameworkId { get; set; }
        [ForeignKey("FrameworkId")]
        public virtual Proficiency Framework { get; set; }

        public virtual IEnumerable<TicketReaction> Reactions { get; set; }

        [ForeignKey("ApplicantId")]
        public virtual DiscordUser Applicant { get; set; }

        public virtual List<TicketUser> Assignees { get; set; }

    }

    public enum TicketStatus
    {
        Open = 0,
        Closed = 1,
        InProgress = 2
    }

    public enum TicketCategory
    {
        DiscordBots = 0,
        Unity = 1,
        Python = 2,
        Web = 3,
        CSharp = 4,
        Undecided = 5
    }

    public class TicketReaction:IEntity
    {
        public int TicketId { get; set; }
        public int FromId { get; set; }
        public int DiscordMessageId { get; set; }

        [ForeignKey("TicketId")]
        public  Ticket Ticket { get; set; }
        
        [ForeignKey("FromId")]
        public DiscordUser From { get; set; }

        [ForeignKey("DiscordMessageId")]
        public DiscordMessage DiscordMessage { get; set; }
    }
}

File: DapperDino.Web/VueApp/ticketWrapper/types/models/ticket.ts
export interface Ticket {
    id: number;
	description:string;
    subject:string;
    applicantId:number;
    priority:number;
    status: TicketStatus;
    createdOn: Date;
    lastModified: Date;
    reactions: TicketReaction[];
    applicant: DiscordUser;
    assignees: TicketUser[];
}

export enum TicketStatus {
    open = 0,
    closed = 1,
    inProgress = 2
}

export enum TicketCategory {
    discordBots = 0,
    unity = 1,
    python = 2,
    web = 3,
    cSharp = 4,
    undecided = 5
}

export interface TicketReaction {
    id: number;
    ticketId: number;
    fromId: number;
    discordMessageId: number;
    ticket: Ticket;
    from: DiscordUser;
    discordMessage: DiscordMessage;
}

export interface TicketUser {

    ticketId: number;
    ticket: Ticket;

    discordUserId: number;
    discordUser: DiscordUser;
}

export interface DiscordUser {
    id: number;
    discordId: string;
    username: string;
    name: string;
    xp: number;
    level: number;
    ticketUsers: TicketUser[];
}

export interface DiscordMessage {
    id: number;
    messageId: string;
    guildId: string;
    channelId: string;
    message: string;
    timestamp: Date;
    isEmbed: boolean;
    isDm: boolean;
    imageLink: string;
    discordUserId: number;
    discordUser: DiscordUser;
}

export interface TicketAction {
    url: string;
    action: string;
}


